INT-4506: Splitter: close stream and Closeables

garyrussell opened the pull request:
JIRA: https://jira.spring.io/browse/INT-4506

Capture the actual splitter result in the `FunctionIterator and close it if necessary.

**cherry-pick to 5.0.x**

garyrussell committed:
INT-4506: Splitter: close stream and Closeables

JIRA: https://jira.spring.io/browse/INT-4506

Capture the actual splitter result in the `FunctionIterator and close it if necessary.

**cherry-pick to 5.0.x**

artembilan commented:
There are some problems with Javadocs:
```
/home/travis/build/spring-projects/spring-integration/spring-integration-core/src/main/java/org/springframework/integration/util/FunctionIterator.java:67: error: @param name not found
	 * @param iterator the iterator.
	          ^
/home/travis/build/spring-projects/spring-integration/spring-integration-core/src/main/java/org/springframework/integration/util/FunctionIterator.java:72: warning: no @param for newIterator
	public FunctionIterator(Iterator<T> newIterator, Function<? super T, ? extends V> function) {
	       ^
/home/travis/build/spring-projects/spring-integration/spring-integration-core/src/main/java/org/springframework/integration/util/FunctionIterator.java:79: error: @param name not found
	 * @param iterator the iterator.
	          ^
/home/travis/build/spring-projects/spring-integration/spring-integration-core/src/main/java/org/springframework/integration/util/FunctionIterator.java:83: warning: no @param for newIterator
	public FunctionIterator(@Nullable Object root, Iterator<T> newIterator, Function<? super T, ? extends V> function) {
	       ^
```

See Travis report.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/util/FunctionIterator.java:
```diff
@@ -64,6 +106,27 @@ public void close() throws IOException {
+			}
+		}
+		if (!this.root.equals(this.iterator)) {
+			if (this.root instanceof Closeable) {
```
See comment in JIRA: https://jira.spring.io/browse/INT-4506?focusedCommentId=161106&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-161106

garyrussell committed:
Add Closeable to proxy in test; remove AutoCloseable check.

Closeable extends AutoCloseable.

garyrussell committed:
Fix javadocs; test for AutoCloseable instead of Closeable

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/util/FunctionIterator.java:
```diff
@@ -64,6 +106,11 @@ public void close() throws IOException {
 			((Closeable) this.iterator).close();
 		}
+		if (!this.root.equals(this.iterator)) {
+			if (this.root instanceof Closeable) {
```
Not enough coffee today 😄

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/util/FunctionIterator.java:
```diff
@@ -61,8 +102,23 @@ public V next() {
+					((AutoCloseable) this.root).close();
+				}
+				catch (Exception e) {
+					// NOSONAR
```
I believe we can change a `CloseableIterator` to the `AutoCloseable` now as well...

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/util/FunctionIterator.java:
```diff
@@ -61,8 +102,23 @@ public V next() {
+			}
+		}
+		if (!this.root.equals(this.iterator)) {
+			if (this.root instanceof AutoCloseable) {
```
Conditions can be combined to a single one.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/util/FunctionIterator.java:
```diff
@@ -16,30 +16,71 @@
+	 * @param function the function.
+	 * @deprecated - use {@link #FunctionIterator(Object, Iterable, Function)}
+	 */
+	@Deprecated
```
Why is that? What is wrong with the regular ctor without a `root`?
Please, see your change into the `XPathMessageSplitter`. Looks like the `root` is not required and there is no reason to enforce end-users to think about the `root` where it isn't necessary.

Plus, let's consider a better Javadocs what is this `root` about here.

And that's why I really would prefer a wrapping to the `CloseableIterator` instead...


Also it's not visible here, but looks like we now can remove `remove()` method: it is a `default` one on the `Iterator`

garyrussell replied:
My concern is this:

```java
public class Foo implements Iterator, Closeable {

...

}

Since we have no control over what the user returns, I believe we **have** to capture the root object so we can close it when the iteration is complete.

I deprecated the other constructors to discourage their use.

But I can see a better way now; will rework...

garyrussell committed:
Polishing - PR comments; tighten up API.

CloseableIterator is now AutoCloseable with overriden close() (no exception)
to avoid lint problem.

artembilan reviewed (approved):
LGTM.
Will merge after green Travis report...

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/util/FunctionIterator.java:
```diff
@@ -16,30 +16,68 @@
+	 * @param function the function.
+	 * @since 5.0.7
+	 */
+	public FunctionIterator(@Nullable AutoCloseable closeable, Iterable<T> iterable,
```
This `closeable` is not propagated to the target ctor.
Will fix on merge.

garyrussell replied:
Doh - thanks.

artembilan commented:
Merged as https://github.com/spring-projects/spring-integration/commit/bebde260db912c4a7065657ae1325150ddd3d047 and cherry-picked to `5.0.x`.
